api/utils/retry_promise.ts

```diff
@@ -0,0 +1,14 @@
+type RetryReturnType = (func: () => Promise<any>, retries: number, error?: Error) => Promise<any>;
+
+export const retryPromise: RetryReturnType = async (func, retries, error) => {
+  if (retries < 0) {
```

Comment:
i think this would be cleaner if we put it inside the `catch` block, and check for when `retries === 0`. that way, you don't need to pass the `Error` down the recursion stack.